components/layout/display_list/mod.rs

```diff
@@ -233,6 +285,149 @@ impl DisplayListBuilder<'_> {
             self.display_list.compositor_info.epoch.as_u16(),
         ))
     }
+
+    /// Draw highlights around the node that is currently hovered in the devtools
+    fn paint_dom_inspector_highlight(&mut self, highlight: HighlightTraversalState) {
+        const CONTENT_BOX_HIGHLIGHT_COLOR: webrender_api::ColorF = webrender_api::ColorF {
+            r: 0.23,
+            g: 0.7,
+            b: 0.87,
+            a: 0.5,
+        };
+
+        const PADDING_BOX_HIGHLIGHT_COLOR: webrender_api::ColorF = webrender_api::ColorF {
+            r: 0.49,
+            g: 0.3,
+            b: 0.7,
+            a: 0.5,
+        };
+
+        const BORDER_BOX_HIGHLIGHT_COLOR: webrender_api::ColorF = webrender_api::ColorF {
+            r: 0.2,
+            g: 0.2,
+            b: 0.2,
+            a: 0.5,
+        };
+
+        const MARGIN_BOX_HIGHLIGHT_COLOR: webrender_api::ColorF = webrender_api::ColorF {
+            r: 1.,
+            g: 0.93,
+            b: 0.,
+            a: 0.5,
+        };
+
+        // Highlight content box
+        let content_box = highlight.content_box.to_webrender();
+        let properties = wr::CommonItemProperties {
+            clip_rect: content_box,
+            spatial_id: highlight.spatial_id,
+            clip_chain_id: highlight.clip_chain_id,
+            flags: wr::PrimitiveFlags::default(),
+        };
+
+        self.display_list
+            .wr
+            .push_rect(&properties, content_box, CONTENT_BOX_HIGHLIGHT_COLOR);
+
+        // Highlight margin, border and padding
+        if let Some((box_fragment, containing_block)) = highlight.maybe_box_fragment {
+            let mut paint_highlight =
+                |color: webrender_api::ColorF,
+                 fragment_relative_bounds: PhysicalRect<Au>,
+                 widths: webrender_api::units::LayoutSideOffsets| {
+                    if widths.is_zero() {
+                        return;
+                    }
+
+                    let bounds = fragment_relative_bounds
+                        .translate(containing_block.origin.to_vector())
+                        .to_webrender();
```

Comment:
This has changed last week, so it makes sense that you didn't know this, but now you can use `BoxFragment::offset_by_containing_block` to do this and you don't need to keep track of the containing block rectangle as part of `maybe_box_fragment`.

```suggestion
                    let bounds = box_fragment
                        .offset_by_containing_block(&fragment_relative_bounds)
                        .to_webrender();
```